### PR TITLE
Etcd v3.1.0-alpha.0 with embedding and gRPC support

### PR DESCRIPTION
**Pull request by smarterclayton**

### CODE REVIEW

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
Flaked on rollout history on deployments.  We appear to be unable to schedule more pods - the service endpoints test, this test, and the deploy iteratively test all rely on having enough room to schedule.  We need to add debugging to it.

**Comment by smarterclayton** [comment]
@liggitt if you can review the etcd master server startup change.

**Comment by smarterclayton** [comment]
```
hack/test-cmd.sh has failed: idle.sh:34: executing 'oc get pod -l app=idling-echo -o go-template='{{ len .items }}'' expecting any result and text '2'; re-trying every 0.2s until completion or 60.000s
``` 

[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
[test]

**Comment by smarterclayton** [comment]
@liggitt review

File: hack/install-etcd.sh [hunk 1]
```diff
@@ -23,10 +23,6 @@ fi
 
 # setup a private GOPATH so the build can succeed
 export GOPATH="${PWD}/gopath"
-rm -f "${GOPATH}/src/github.com/coreos/etcd"
-mkdir -p "${GOPATH}/src/github.com/coreos"
-ln -s "${PWD}" "${GOPATH}/src/github.com/coreos/etcd"
```
Comment (liggitt) [inline hack/install-etcd.sh]: what was the original reason for this, and why is it no longer needed?

File: test/cmd/basicresources.sh [hunk 2]
```diff
@@ -26,7 +26,8 @@ os::test::junit::declare_suite_start "cmd/basicresources/versionreporting"
 os::build::get_version_vars
 os_git_regex="$( escape_regex "${OS_GIT_VERSION%%-*}" )"
 kube_git_regex="$( escape_regex "${KUBE_GIT_VERSION%%-*}" )"
-etcd_git_regex="$( escape_regex "${ETCD_GIT_VERSION%%-*}" )"
+etcd_version="$(echo "${ETCD_GIT_VERSION}" | sed -E "s/\-.*//g" | sed -E "s/v//")"
+etcd_git_regex="$( escape_regex "${etcd_version%%-*}" )"
```
Comment (liggitt) [inline test/cmd/basicresources.sh]: *raises eyebrow*

File: pkg/cmd/server/etcd/etcdserver/run.go [hunk 3]
```diff
@@ -1,64 +1,95 @@
...
 	}
+
 	go func() {
-		glog.Infof("Started etcd at %s", etcdServerConfig.Address)
-		<-stopped
+		defer e.Close()
+
+		select {
+		case <-e.Server.ReadyNotify():
+			glog.Infof("Started etcd at %s", etcdServerConfig.Address)
```
Comment (liggitt) [inline pkg/cmd/server/etcd/etcdserver/run.go]: this seems wrong... if we reach this log line, we'll immediately fatal below?

File: pkg/cmd/server/etcd/etcdserver/run.go [hunk 4]
```diff
@@ -1,64 +1,95 @@
...
+			glog.Warning("etcd took too long to start, stopped")
+			e.Server.Stop() // trigger a shutdown
+		}
+		glog.Fatalf("etcd has returned an error: %v", <-e.Err())
 	}()
 }
+
+// addressToURLs turns a host:port comma delimited list into an array valid
+// URL strings with the appropriate prefix for the TLS mode.
+func addressToURLs(addr string, isTLS bool) []string {
```
Comment (liggitt) [inline pkg/cmd/server/etcd/etcdserver/run.go]: we're passing in bind addresses here... is `https://0.0.0.0:4001` going to be valid?

File: pkg/cmd/server/etcd/etcdserver/server.go [hunk 5]
```diff
@@ -1,255 +0,0 @@
...
-		PeerTLSInfo: cfg.peerTLSInfo,
-	}
-	var s *etcdserver.EtcdServer
-	s, err = etcdserver.NewServer(srvcfg)
-	if err != nil {
-		return nil, err
-	}
-	osutil.HandleInterrupts()
-	s.Start()
-	osutil.RegisterInterruptHandler(s.Stop)
```
Comment (liggitt) [inline pkg/cmd/server/etcd/etcdserver/server.go]: were the interrupt handlers important?

**Comment by liggitt** [comment]
if someone starts an existing data store with this, it'll be impossible for them to roll back to an older version of origin, right? if so, worth noting in the release notes

File: hack/install-etcd.sh (hunk 1 above)
Comment (smarterclayton) [inline hack/install-etcd.sh]: etcd ./build does this now - they didn't used to control their gopath, now they do

File: test/cmd/basicresources.sh (hunk 2 above)
Comment (smarterclayton) [inline test/cmd/basicresources.sh]: etcd doesn't follow our regex rules

**Comment by smarterclayton** [comment]
Impossible if they start older than 2.3.7 I believe.

File: pkg/cmd/server/etcd/etcdserver/run.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/cmd/server/etcd/etcdserver/run.go]: Yes, same as before.  The logic for calculating addresses simply moved deeper.

File: pkg/cmd/server/etcd/etcdserver/server.go (hunk 5 above)
Comment (smarterclayton) [inline pkg/cmd/server/etcd/etcdserver/server.go]: Yes, added them back.

Comment (smarterclayton) [inline pkg/cmd/server/etcd/etcdserver/server.go]: Although now it's survivable.

File: pkg/cmd/server/etcd/etcdserver/run.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/cmd/server/etcd/etcdserver/run.go]: It waits on the error channel.

File: test/cmd/basicresources.sh (hunk 2 above)
Comment (smarterclayton) [inline test/cmd/basicresources.sh]: follow our version style rules

File: pkg/cmd/server/etcd/etcdserver/run.go (hunk 3 above)
Comment (liggitt) [inline pkg/cmd/server/etcd/etcdserver/run.go]: ah

File: pkg/cmd/server/etcd/etcdserver/run.go [hunk 6]
```diff
@@ -1,64 +1,95 @@
...
+	cfg.Name = defaultName
+	cfg.Dir = etcdServerConfig.StorageDir
 
-		initialClusterToken: "etcd-cluster",
+	clientTLS := configapi.UseTLS(etcdServerConfig.ServingInfo)
+	if clientTLS {
+		cfg.ClientTLSInfo.CAFile = etcdServerConfig.ServingInfo.ClientCA
+		cfg.ClientTLSInfo.CertFile = etcdServerConfig.ServingInfo.ServerCert.CertFile
+		cfg.ClientTLSInfo.KeyFile = etcdServerConfig.ServingInfo.ServerCert.KeyFile
+		cfg.ClientTLSInfo.ClientCertAuth = len(cfg.ClientTLSInfo.CertFile) > 0
```
Comment (liggitt) [inline pkg/cmd/server/etcd/etcdserver/run.go]: this should trigger off the ClientCA being specified, not the cert file

File: pkg/cmd/server/etcd/etcdserver/run.go [hunk 7]
```diff
@@ -1,64 +1,95 @@
...
-	if configapi.UseTLS(etcdServerConfig.PeerServingInfo) {
-		cfg.peerTLSInfo.CAFile = etcdServerConfig.PeerServingInfo.ClientCA
-		cfg.peerTLSInfo.CertFile = etcdServerConfig.PeerServingInfo.ServerCert.CertFile
-		cfg.peerTLSInfo.KeyFile = etcdServerConfig.PeerServingInfo.ServerCert.KeyFile
+	peerTLS := configapi.UseTLS(etcdServerConfig.PeerServingInfo)
+	if peerTLS {
+		cfg.PeerTLSInfo.CAFile = etcdServerConfig.PeerServingInfo.ClientCA
+		cfg.PeerTLSInfo.CertFile = etcdServerConfig.PeerServingInfo.ServerCert.CertFile
+		cfg.PeerTLSInfo.KeyFile = etcdServerConfig.PeerServingInfo.ServerCert.KeyFile
+		cfg.PeerTLSInfo.ClientCertAuth = len(cfg.PeerTLSInfo.CertFile) > 0
```
Comment (liggitt) [inline pkg/cmd/server/etcd/etcdserver/run.go]: same here, len(ClientCA) > 0

**Comment by smarterclayton** [comment]
[test] idle.sh flake

**Comment by liggitt** [comment]
LGTM after squash of review commit

**Comment by smarterclayton** [comment]
[merge]

**Comment by smarterclayton** [comment]
[test] #11024

**Pull request merged**